### PR TITLE
HPCC-29549 Handle fully qualified path when validating DZ scope access

### PR DESCRIPTION
**Pull request by wangkx**
1. Move the code of the scope permission validation from the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to the readAndCheckSpraySourceReq().
2. In readAndCheckSpraySourceReq(), parse the file requests inside the SourcePath request. For each file request, validating the scope permission. If a file request contains a host/ip, validate the host/ip and local file path. If a file request does not contain a host/ip, validate the file path.
3. Change the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to setSpraySourceDFUFileSpec().
4. Do not validate DZ scope if no DZ configured and no DropZoneRestriction.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
Noticed 1 smoketest failure: Spray variable with wrong queue name: Fail. But, I do not know why the queue name is wrong.

**Comment by wangkx** [comment]
@jakesmith a smoketest failed. I will take a look.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 1]
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
-            sourcePlaneReq.append(srcPlane);
         }
         else
         {
-            sourceIPReq.set(srcIP).trim();
-            if (sourceIPReq.isEmpty())
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            sourcePlaneReq.append(plane->queryProp("@name"));
+            if (firstFileReq)
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1982]: could this be if (nullptr == dropZone) ?

And, it should be an error in cloud and in bare-metal unless isDropZoneRestrictionEnabled() = false if the dropzone cannot be found.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 2]
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+            {
+                firstFileReq = false;
+
+                //Use the fileReq and sourceHostReq to find out dropzone plane.
+                dropZone.setown(getAndValidateDropZone(fileReq, sourceHostReq));
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1987]: it feels this should proceed the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978).
It also might make this method clear, if that code were in a separate (static) helper function (can still be within ws_fsService.cpp).

**Review by jakesmith — COMMENTED** [review]
@wangkx - I think the consolidation into readAndCheckSpraySourceReq and checking of the UNC paths is ok, but the code is quite hard to follow at the moment, and I am not sure the order things are validated is correct.
I think it could be re-organised and made clearer. See comments.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 3]
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+
+
+    SocketEndpoint srcHostEp;
+    if (!sourceHostReq.isEmpty())
+        srcHostEp.set(sourceHostReq);
+
+    StringBuffer sourcePath, s(srcPath);
+    s.trim();
+    if (s.isEmpty())
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source path not specified.");
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:1921]: Will change

File: esp/services/ws_fs/ws_fsService.cpp [hunk 4]
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
-void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char* srcIP, const char* srcPath, const char* srcPlane,
-    StringBuffer& sourcePlaneReq, StringBuffer& sourceIPReq, StringBuffer& sourcePathReq)
+void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, MemoryBuffer& srcxml, const char* srcHost, const char* srcPath, const char* srcPlane,
+    StringBuffer& sourcePlaneReq, StringBuffer& sourceHostReq, StringBuffer& sourcePathReq)
 {
-    StringBuffer sourcePath(srcPath);
-    sourcePath.trim();
-    if(srcxml.length() == 0)
+    if(srcxml.length() > 0)
+        return;
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:1888]: Will change

File: esp/services/ws_fs/ws_fsService.cpp [hunk 5]
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+            continue;
+
+        //Find and check possible host/ip inside the fileReq.
+        const char* hostInPath = nullptr;
+        StringBuffer localPath, hostInPathStr;
+        if (isPathSepChar(fileReq[0]) && (fileReq[0] == fileReq[1]))
+        {
+            splitUNCFilename(fileReq, &hostInPathStr, &localPath, &localPath, &localPath);
+            const char* hostInPath = hostInPathStr.str() + 2;
+            if (!isEmptyString(srcPlane))
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:1940]: Will change

File: esp/services/ws_fs/ws_fsService.cpp (hunk 2 above)
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:1987]: If this proceeds the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978), the getAndValidateDropZone() will be called when the Source Plan is not empty.
I am not sure we want to call the getAndValidateDropZone() when the Source Plane is not empty. When the Source Plane is not empty, the sourceHostReq has been validated and the fileReq will be validated inside the  if (isAbsolutePath(fileReq)) block.

**Comment by wangkx** [comment]
> the code is quite hard to follow at the moment,
I am going to re-organize the code to see if it will be better.

**Comment by wangkx** [comment]
@jakesmith I did the changes and re-organized the code. Please review.

**Comment by jakesmith** [comment]
@wangkx - PR and commit message should be "Handle fully qualified .. "

File: esp/services/ws_fs/ws_fsService.cpp [hunk 6]
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+
+    //Clean srcPath
+    StringBuffer sourcePath, s(srcPath);
+    s.trim();
+    if (s.isEmpty())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "Source path for spraying a file not specified.");
+    if (!sourcePlaneReq.isEmpty())
+        getStandardPosixPath(sourcePath, s.str());
+    else
+        getStandardPosixPath(sourcePathReq, s.str());
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1932]: sorry, I don't understand the logic here..
If no plane in request, then fill incoming srcPath, into [outbound] sourcePathReq, leaving 'sourcePath' empty,
which is then used below (line 1935)..

Needs comments to clarify at least.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 7]
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
-                addNonEmptyPathSepChar(sourcePath);
-                sourcePath.append(s);
-            }
-            getDropZoneHost(srcPlane, dropZone, sourceIPReq);
-            sourcePlaneReq.append(srcPlane);
+    if (!isPathSepChar(fileReq[0]) | (fileReq[0] != fileReq[1]))
+        return;
+
+    splitUNCFilename(fileReq, &hostInPathStr, &localPath, &localPath, &localPath);
+    const char* hostInPath = hostInPathStr.str() + 2;
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1890]: needs/using bit after '//', i.e. the <hostname>, but returning (in hostInPathStr) the //<hostname> ..
that is confusing.
Should return the hostname - that if anything will be whats needed, caller should have to do ... + 2

Code clarity here is far more important than a slight gain in efficiency because need to copy in to output StringBuffer.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 8]
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+        const char* fileReq = fileReqs.item(i);
+        if (isEmptyString(fileReq))
+            continue;
+
+        //Parse the fileReq to find/validate possible host/ip and local path.
+        StringBuffer hostInPath, localPath, absPath;
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath);
+
+        //Validate file path.
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1970]: would be clearer if checkHostInSpraySourcePath had returned true if was a UNC path and had validated, and do, e.g.:
```
const char *path;
if (IsHostInSprayPath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath))
    path = localPath.str();
else
   path = fileReq;

File: esp/services/ws_fs/ws_fsService.cpp [hunk 9]
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "Source path for spraying a file not specified.");
+    if (!sourcePlaneReq.isEmpty())
+        getStandardPosixPath(sourcePath, s.str());
+    else
+        getStandardPosixPath(sourcePathReq, s.str());
+
+    StringArray fileReqs;
+    fileReqs.appendList(sourcePath, ","); // handles comma separated files
+
+    if (sourcePlaneReq.isEmpty())
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1937]: let us try to separate and label code like this as much as possible to emphasis it's legacy.
A comment here and above the definition of 'checkSpraySourcePathAndGetPlane' will suffice.

It will make future reading easier to follow the standard/normal/expected code path.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 10]
```diff
@@ -2008,7 +2118,7 @@ bool CFileSprayEx::onSprayFixed(IEspContext &context, IEspSprayFixed &req, IEspS
         wu->setCommand(DFUcmd_import);
 
         IDFUfileSpec *source = wu->queryUpdateSource();
-        checkDZScopeAccessAndSetSpraySourceDFUFileSpec(context, sourcePlaneReq, sourceIPReq, srcfile, srcxml, source);
+        setSpraySourceDFUFileSpec(sourceHostReq, sourcePathReq, srcxml, source);
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:2121]: This is probably going to need revising to ensure it includes plane name, following this change: https://github.com/hpcc-systems/HPCC-Platform/pull/17353/files#diff-8b47ea38bf0ea671d5db7b9ae0166a6c19a1ad525011ffac7556726d8245844a

And for that reason and others this PR should be rebased onto master.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 11]
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
...
+        StringBuffer hostInPath, localPath;
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHost, &sourceHostEp, localPath, hostInPath);
+
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
+        if (containsRelPaths(path)) //Detect a path like: a/../../../f
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", fileReq);
+
+        if (firstFileReq)
+        {
+            firstFileReq = false;
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:2025]: still seems to a logical problem here..
What happens if 2nd, 3rd UNC path have different hostname in them?

Looks like 1st will be validated, and all others ignored at the moment?
But judging by setSpraySourceDFUFileSpec and srcDFUfileSpec->setMultiFilename .. we don't support it anyway.. i.e. only 1 host has ever been permitted here.

If so, this should not ignore other UNC with different hosts, but throw an error and reject it

File: esp/services/ws_fs/ws_fsService.cpp [hunk 12]
```diff
@@ -2133,9 +2243,9 @@ bool CFileSprayEx::onSprayVariable(IEspContext &context, IEspSprayVariable &req,
             gName.append(destNodeGroup);
 
         MemoryBuffer& srcxml = (MemoryBuffer&)req.getSrcxml();
-        StringBuffer sourcePlaneReq, sourceIPReq, sourcePathReq;
-        readAndCheckSpraySourceReq(srcxml, req.getSourceIP(), req.getSourcePath(), req.getSourcePlane(), sourcePlaneReq, sourceIPReq, sourcePathReq);
-        const char* srcfile = sourcePathReq.str();
+        StringBuffer sourcePlaneReq, sourceHostReq, sourcePathReq;
+        if (srcxml.length() == 0)
+            readAndCheckSpraySourceReq(context, req.getSourceIP(), req.getSourcePath(), req.getSourcePlane(), sourcePlaneReq, sourceHostReq, sourcePathReq);
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:2248]: not for now: there's a lot in common with onSprayFixed vs onSprayVariable, but differences that I can't immediately see why are there, e.g. onSprayVariable has:
```
        StringBuffer gName, ipAddr;
        const char *pTr = strchr(destNodeGroup, ' ');
        if (pTr)
        {
            gName.append(pTr - destNodeGroup, destNodeGroup);
            ipAddr.append(pTr+1);
        }
        else
            gName.append(destNodeGroup);
```

They should be as similar as possible, preferably using common code wherever possible.
Not for now, but could the ecm EspSpray*'s be commoned up and the *Fixed vs *Variable definitions be derived from EspSprayCommon? Such that these methods wouldn't have to duplicate as much?

**Review by jakesmith — COMMENTED** [review]
@wangkx - I think it's clearer, but still quite quite convoluted and hard to follow the 'normal' code path vs the edge/exception/legacy cases.
To some extent this is unavoidable because we are having to be backward compatible and support weird cases.
Please see comments.

File: esp/services/ws_fs/ws_fsService.cpp (hunk 11 above)
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:2025]: I do assume all of the UNC paths have the same hostname now. If there are different hostnames, could they are in different planes? If so, which one should be set into the DFU WU or we throw an error?

Comment (jakesmith) [reply esp/services/ws_fs/ws_fsService.cpp:2025]: as I say, afaics, srcDFUfileSpec->setMultiFilename etc. wouldn't support it anyway, so should throw an error.

File: esp/services/ws_fs/ws_fsService.cpp (hunk 10 above)
Comment (jakesmith) [reply esp/services/ws_fs/ws_fsService.cpp:2121]: >Not sure why we need to 'ensure it includes plane name'? Do we want to call the source->setGroupName(sourcePlaneReq); inside the setSpraySourceDFUFileSpec()?

See https://track.hpccsystems.com/browse/HPCC-29553
It ensures the group names are created (in bare-metal) for planes, meaning the logical files (IFileDescriptor's) will have properly formed group not IP's within them.
The IFileDescriptor's created by DFU need to know the group(plane) name, and hence this (setGroupName) is the means by which they do.

Comment (jakesmith) [reply esp/services/ws_fs/ws_fsService.cpp:2121]: >If this PR is rebased onto master, should we create another small 9.0.x PR to fix the crash when isDropZoneRestrictionEnabled = false and getAndValidateDropZone() returns nullptr and to not call getDZFileScopePermissions() if no plane name is found?

Yes, that makes sense.

**Comment by wangkx** [comment]
Will add more changes after https://github.com/hpcc-systems/HPCC-Platform/pull/17378.

File: esp/services/ws_fs/ws_fsService.cpp (hunk 11 above)
Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:2025]: I just realized that the checkHostInSpraySourcePath() does throw an error if hostnames are from different hosts. If a dropzone is found for the first file, the dropzone will be used for all files.

Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp:2025]: The code should be revised to check the path also if a dropzone is found: if (!startsWith(path, dropZonePlanePath)) ...

**Comment by wangkx** [comment]
I am closing this PR for now. Will reopen after https://github.com/hpcc-systems/HPCC-Platform/pull/17378.
